quasar-utils/src/main/groovy/net/corda/plugins/quasar/QuasarPlugin.groovy

```diff
@@ -100,6 +130,14 @@ class QuasarPlugin implements Plugin<Project> {
                         "-Dco.paralleluniverse.fibers.verifyInstrumentation"
             }
         }
+        project.tasks.withType(JavaCompile).configureEach {
+            doFirst {
+                if(extension.enableJavacPlugin.get()) {
+                    options.annotationProcessorPath += project.configurations[JAVAC_PLUGIN]
+                    options.compilerArgs += "-Xplugin:net.corda.plugins.javac.quasar.SuspendableChecker"
+                }
+            }
+        }
```

Comment:
How does this work for Kotlin classes?

Reply:
unfortunately this only works for Java code (but suspendable methods defined in Kotlin should be detected correctly, it would be good to add a test for it though)

Reply:
I don't think that "only works for Java code" is good enough when both Corda and our CorDapp API are written in Kotlin. I suspect that it would be better not to use any compiler plugins at all, but to scan all the compiled `.class` files as a separate task that finalizes `classes`.

This task would need to take account of the configurable `@Suspendable` annotation too.

Reply:
I will add a way to configure the name of the `@Suspendable` annotation. Regarding kotlin, the vast majority of our customers use Java, so this plugin will be useful for them, there is apparently [a way](https://github.com/JetBrains/kotlin/blob/master/plugins/allopen/allopen-cli/src/AllOpenPlugin.kt) to do the same thing for `kotlinc`. I believe a compiler plugin gives an overall better user experience:

- class files only contain line number information, this may be less clear with code that contains very long lines
- class files don't know where the file exactly is on the filesystem, only contain
- a separate verification check won't be automatically invoked by the IDE when building the project and is likely to only be invoked on CI pipeline, as it currently happens with `detekt`

Reply:
Unfortunately, Kotlin is a "first class citizen" and so I don't believe that we can support a feature like this for Java only, but I can obviously raise this with @mnesbit and @fenryka in the New Year to see what they think.

As for your claim about "the vast majority of our customers", I can only point out that the vast majority of the issues that have been escalated to me from Support have used Kotlin rather than Java :expressionless:.

If your goal here is to provide developer feedback via the IDE when why not implement this as an IntelliJ plugin instead?

Reply:
I am aware of Kotlin compiler plugins too, but these seem to be strongly coupled to the particular version of Kotlin being used. And I have fought _tooth and nail_ over the years to prevent `quasar-utils` becoming responsible for configuring _all_ versions of _all_ compilers for _every_ JVM language that someone could conceivably write a CorDapp in.

I don't know if you've ever watched "The Twilight Zone", but there was an episode called ["The Curious Case of Edgar Witherspoon"](https://en.wikipedia.org/wiki/The_Curious_Case_of_Edgar_Witherspoon) that is _eerily close_ as an analogy for how the Corda Gradle plugins will function in Corda some day, unless we're careful.